python/lancedb/query.py

```diff
@@ -54,46 +55,115 @@ class Query(pydantic.BaseModel):
     refine_factor: Optional[int] = None
 
 
-class LanceQueryBuilder:
-    """
-    A builder for nearest neighbor queries for LanceDB.
+class LanceQueryBuilder(ABC):
+    @classmethod
+    def create(
+        cls,
+        table: "lancedb.table.Table",
+        query: Optional[Union[np.ndarray, str]],
+        query_type: str,
+        vector_column_name: str,
+    ) -> LanceQueryBuilder:
+        if query is None:
+            return LanceEmptyQueryBuilder(table)
+
+        query, query_type = cls._validate_query(
+            table, query, query_type, vector_column_name
+        )
 
-    Examples
-    --------
-    >>> import lancedb
-    >>> data = [{"vector": [1.1, 1.2], "b": 2},
-    ...         {"vector": [0.5, 1.3], "b": 4},
-    ...         {"vector": [0.4, 0.4], "b": 6},
-    ...         {"vector": [0.4, 0.4], "b": 10}]
-    >>> db = lancedb.connect("./.lancedb")
-    >>> table = db.create_table("my_table", data=data)
-    >>> (table.search([0.4, 0.4])
-    ...       .metric("cosine")
-    ...       .where("b < 10")
-    ...       .select(["b"])
-    ...       .limit(2)
-    ...       .to_df())
-       b      vector  _distance
-    0  6  [0.4, 0.4]        0.0
-    """
+        if isinstance(query, str):
+            # fts
+            return LanceFtsQueryBuilder(table, query)
```

Comment:
The docstring said:

> If the query is a string, then the query type is "vector" if the
> table has embedding functions else the query type is "fts"

But I don't see that reflected here. Am I missing something?

Reply:
validate_query can change the query type . Should be renamed resolve_query type probably

Reply:
Ah that makes sense. Getting so used to Rust where we get to know what parameters are mutated and which aren't :)